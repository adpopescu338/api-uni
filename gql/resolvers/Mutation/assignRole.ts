import { Context } from 'libs/types';
import { MutationAssignRoleArgs } from 'libs/types/generated';
import { throwIfNotAdmin } from '../../../libs/throwIfNotAdmin';

export const assignRole = async (
  _: unknown,
  { userId, roleId }: MutationAssignRoleArgs,
  { prisma, user }: Context,
) => {
  throwIfNotAdmin(user);

  const result = await prisma.user.update({
    where: {
      id: userId,
    },
    data: {
      roles: {
        connect: {
          id: roleId,
        },
      },
    },
    include: {
      roles: {
        include: {
          permissions: true,
        },
      },
    },
  });

  return result;
};
