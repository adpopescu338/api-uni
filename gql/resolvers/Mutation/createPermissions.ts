import { Context } from 'libs/types';
import { MutationCreatePermissionsArgs } from 'libs/types/generated';

export const createPermissions = async (
  _: unknown,
  { permissions }: MutationCreatePermissionsArgs,
  { prisma }: Context,
) => {
  const createdPermissions = await prisma.permission.createMany({
    data: permissions,
  });

  return createdPermissions;
};
