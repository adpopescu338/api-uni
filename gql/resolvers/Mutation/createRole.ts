import { Context } from 'libs/types';
import { MutationCreateRoleArgs } from 'libs/types/generated';
import { throwIfNotAdmin } from 'libs/throwIfNotAdmin';

export const createRole = async (
  _: unknown,
  { name, description, permissionIds }: MutationCreateRoleArgs,
  { prisma, user }: Context,
) => {
  throwIfNotAdmin(user);

  const role = await prisma.userDefinedRole.create({
    data: {
      name,
      description,
      ...(permissionIds.length && {
        permissions: {
          connect: permissionIds.map((id) => ({ id })),
        },
      }),
    },
  });

  return role;
};
