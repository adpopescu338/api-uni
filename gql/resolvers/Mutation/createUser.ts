import { Context } from 'libs/types';
import { MutationCreateUserArgs } from 'libs/types/generated';
import { throwIfNotAdmin } from '../../../libs/throwIfNotAdmin';
import { getHashedPassword } from '../../../libs/getHashedPassword';

export const createUser = async (
  _: unknown,
  { input }: MutationCreateUserArgs,
  { prisma, user, cacheManager }: Context,
) => {
  throwIfNotAdmin(user);

  const password = await getHashedPassword(input.password);

  const newUser = await prisma.user.create({
    data: {
      ...input,
      password,
      isSysAdmin: input.isSysAdmin || false,
    },
  });

  cacheManager.del(cacheManager.cacheKeys.usersCount);

  return {
    ...newUser,
    roles: [],
  };
};
