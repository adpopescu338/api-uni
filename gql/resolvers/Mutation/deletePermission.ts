import { Context } from 'libs/types';
import { MutationDeletePermissionArgs } from 'libs/types/generated';
import { throwIfNotAdmin } from '../../../libs/throwIfNotAdmin';
import { GraphQLError } from 'graphql/error/GraphQLError';

export const deletePermission = async (
  _: unknown,
  { id }: MutationDeletePermissionArgs,
  { prisma, user }: Context,
) => {
  throwIfNotAdmin(user);

  const permission = await prisma.permission.delete({
    where: {
      id,
    },
  });

  if (!permission) {
    throw new GraphQLError('Permission not found', {
      extensions: {
        code: 'PERMISSION_NOT_FOUND',
      },
    });
  }

  return { success: true };
};
