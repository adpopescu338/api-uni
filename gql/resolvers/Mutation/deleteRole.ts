import { Context } from 'libs/types';
import { MutationDeleteRoleArgs } from 'libs/types/generated';
import { throwIfNotAdmin } from 'libs/throwIfNotAdmin';
import { GraphQLError } from 'graphql';

export const deleteRole = async (
  _: unknown,
  { id }: MutationDeleteRoleArgs,
  { prisma, user }: Context,
) => {
  throwIfNotAdmin(user);

  const result = await prisma.userDefinedRole.delete({
    where: {
      id,
    },
  });

  if (!result) {
    throw new GraphQLError('Role not found', {
      extensions: {
        code: 'ROLE_NOT_FOUND',
      },
    });
  }

  return { success: true };
};
