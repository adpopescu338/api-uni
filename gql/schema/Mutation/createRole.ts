import { gql } from 'graphql-tag';

export const createRole = gql`
  type Mutation {
    """
    Only sysadmins can create roles
    """
    createRole(description: String, name: String!, permissionIds: [ID!]!): UserDefinedRole!
  }
`;
