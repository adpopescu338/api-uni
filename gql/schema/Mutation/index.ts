import { logout } from './logout';
import { login } from './login';

import { createRole } from './createRole';
import { updateRole } from './updateRole';
import { deleteRole } from './deleteRole';
import { assignRole } from './assignRole';

import { createPermissions } from './createPermissions';
import { updatePermissions } from './updatePermissions';
import { deletePermission } from './deletePermission';

import { createUser } from './createUser';
import { updateUser } from './updateUser';

export const Mutation = [
  login,
  logout,

  createRole,
  updateRole,
  deleteRole,
  assignRole,

  createPermissions,
  updatePermissions,
  deletePermission,

  createUser,
  updateUser,
];
