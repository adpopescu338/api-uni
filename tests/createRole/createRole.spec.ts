import { describe, it, expect, jest } from '@jest/globals';
import { createRole } from '../../gql/resolvers/Mutation/createRole';

const payload = {
  name: 'name',
  description: 'description',
  permissionIds: ['id'],
};

describe('createRole', () => {
  it('Should be defined', () => {
    expect(createRole).toBeDefined();
  });

  it('Should throw an error if user is not system admin', async () => {
    await expect(
      createRole(undefined, payload, { user: { isSysAdmin: false } } as any),
    ).rejects.toThrowError();
  });

  it('Should save the correct data and return the role', async () => {
    const expectedResult = {
      id: 'id',
      name: 'name',
      description: 'description',
    };

    const ctx = {
      prisma: {
        userDefinedRole: {
          create: jest.fn().mockResolvedValue(expectedResult as never),
        },
      },
      user: {
        isSysAdmin: true,
      },
    } as any;

    const result = await createRole(undefined, payload, ctx);
    const { lastCall } = ctx.prisma.userDefinedRole.create.mock;

    expect(result).toEqual(expectedResult);
    expect(lastCall).toMatchSnapshot();
  });
});
