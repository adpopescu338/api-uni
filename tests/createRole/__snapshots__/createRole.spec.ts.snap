// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createRole Should save the correct data and return the role 1`] = `
[
  {
    "data": {
      "description": "description",
      "name": "name",
      "permissions": {
        "connect": [
          {
            "id": "id",
          },
        ],
      },
    },
  },
]
`;
